### PR TITLE
syncer: fix safemode exit too early

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve? 
as title, since many events share a same GTID, the checking of exitSafeMode should not use `equal` otherwise some partitially replicated GTID will cause duplicate entry error.

### What is changed and how it works?

we now make sure the replication progress is later (not equal) than exitSafeModePoint when exit safe-mode

### Check List 

Tests 

 - Integration test

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

File: syncer/syncer.go [hunk 1]
```diff
@@ -1305,6 +1305,13 @@ func (s *Syncer) syncDML(
 			}
 		})
 
+		failpoint.Inject("SafeModeExit", func(val failpoint.Value) {
+			if intVal, ok := val.(int); ok && intVal == 4 {
+				s.tctx.L().Warn("fail to exec DML", zap.String("failpoint", "SafeModeExit"))
+				failpoint.Return(0, terror.ErrDBExecuteFailed.Delegate(errors.New("SafeModeExit"), "mock"))
+			}
+		})
+
```
Comment (lichunzhu) [inline syncer/syncer.go]: Actually I want to `executeSQLs` and then quit here. Directly return error here can't explain safe mode does take effect.

File: syncer/syncer.go [hunk 2]
```diff
@@ -1358,6 +1351,13 @@ func (s *Syncer) syncDML(
 
 			if idx >= count || sqlJob.tp == flush {
 				affect, err = executeSQLs()
+
+				failpoint.Inject("SafeModeExit", func(val failpoint.Value) {
```
Comment (glorv) [inline syncer/syncer.go:1355]: Shall we put this failpoint into `executeSQLs`

Comment (lance6716) [reply syncer/syncer.go:1355]: take a look at the above comment in "conversation" tab

and I'll add the length check of `job` to fix panic soon

Comment (lance6716) [reply syncer/syncer.go]: done

**Comment by lance6716** [comment]
don't know why `CreateSchemaIfNotExists` panic ...

```
[2021-07-28T06:56:33.407Z] github.com/pingcap/goleveldb/leveldb.(*DB).newRawIterator(0xc000c9a1a0, 0x0, 0x0, 0x0, 0x0, 0xc0030bc870, 0x0, 0x0, 0x0)
[2021-07-28T06:56:33.407Z] 	/nfs/cache/mod/github.com/pingcap/goleveldb@v0.0.0-20191226122134-f82aafb29989/leveldb/db_iter.go:54 +0xa41
[2021-07-28T06:56:33.407Z] github.com/pingcap/goleveldb/leveldb.(*DB).newIterator(0xc000c9a1a0, 0x0, 0x0, 0x0, 0x0, 0x125a, 0xc0018d7c58, 0x0, 0x0)
[2021-07-28T06:56:33.407Z] 	/nfs/cache/mod/github.com/pingcap/goleveldb@v0.0.0-20191226122134-f82aafb29989/leveldb/db_iter.go:79 +0x506
[2021-07-28T06:56:33.407Z] github.com/pingcap/goleveldb/leveldb.(*DB).NewIterator(0xc000c9a1a0, 0xc0018d7c58, 0x0, 0x0, 0x0)
[2021-07-28T06:56:33.407Z] 	/nfs/cache/mod/github.com/pingcap/goleveldb@v0.0.0-20191226122134-f82aafb29989/leveldb/db.go:884 +0x296
[2021-07-28T06:56:33.407Z] github.com/pingcap/tidb/store/mockstore/mocktikv.newIterator(0xc000c9a1a0, 0xc0018d7c58, 0x0)
[2021-07-28T06:56:33.408Z] 	/nfs/cache/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20210330094614-60111e1c4b6f/store/mockstore/mocktikv/mvcc_leveldb.go:148 +0x95
[2021-07-28T06:56:33.408Z] github.com/pingcap/tidb/store/mockstore/mocktikv.prewriteMutation(0xc000c9a1a0, 0xc0006b2280, 0xc00098e3c0, 0x5ebaf94add80000, 0xc000aa9023, 0x1b, 0x1b, 0xbbb, 0x1, 0x0, ...)
[2021-07-28T06:56:33.408Z] 	/nfs/cache/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20210330094614-60111e1c4b6f/store/mockstore/mocktikv/mvcc_leveldb.go:773 +0x228
[2021-07-28T06:56:33.408Z] github.com/pingcap/tidb/store/mockstore/mocktikv.(*MVCCLevelDB).Prewrite(0xc000f8e570, 0xc0021e6a90, 0x0, 0x0, 0x0)
[2021-07-28T06:56:33.408Z] 	/nfs/cache/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20210330094614-60111e1c4b6f/store/mockstore/mocktikv/mvcc_leveldb.go:675 +0xdd9
[2021-07-28T06:56:33.408Z] github.com/pingcap/tidb/store/mockstore/mocktikv.kvHandler.handleKvPrewrite(0xc0001b6320, 0xc0021e6a90, 0x0)
[2021-07-28T06:56:33.408Z] 	/nfs/cache/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20210330094614-60111e1c4b6f/store/mockstore/mocktikv/rpc.go:205 +0x2bd
[2021-07-28T06:56:33.408Z] github.com/pingcap/tidb/store/mockstore/mocktikv.(*RPCClient).SendRequest(0xc0008d40a0, 0x84f9d08, 0xc00021e010, 0xc00143e020, 0x6, 0xc0021e6b60, 0x4a817c800, 0x0, 0x0, 0x0)
[2021-07-28T06:56:33.408Z] 	/nfs/cache/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20210330094614-60111e1c4b6f/store/mockstore/mocktikv/rpc.go:749 +0x1656
[2021-07-28T06:56:33.408Z] github.com/pingcap/tidb/store/tikv.reqCollapse.SendRequest(0x84ddb30, 0xc0008d40a0, 0x84f9d08, 0xc00021e010, 0xc00143e020, 0x6, 0xc0021e6b60, 0x4a817c800, 0x0, 0x0, ...)
[2021-07-28T06:56:33.408Z] 	/nfs/cache/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20210330094614-60111e1c4b6f/store/tikv/client_collapse.go:49 +0x2f4
[2021-07-28T06:56:33.408Z] github.com/pingcap/tidb/store/tikv.(*RegionRequestSender).sendReqToRegion(0xc000e0c280, 0xc000156400, 0xc000156480, 0xc0021e6b60, 0x4a817c800, 0x0, 0x0, 0x0, 0x0)
[2021-07-28T06:56:33.408Z] 	/nfs/cache/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20210330094614-60111e1c4b6f/store/tikv/region_request.go:441 +0x16cf
[2021-07-28T06:56:33.408Z] github.com/pingcap/tidb/store/tikv.(*RegionRequestSender).SendReqCtx(0xc000e0c280, 0xc000156400, 0xc0021e6b60, 0x3, 0x0, 0x0, 0x4a817c800, 0x7fe134296d00, 0x0, 0x0, ...)
[2021-07-28T06:56:33.408Z] 	/nfs/cache/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20210330094614-60111e1c4b6f/store/tikv/region_request.go:298 +0x133e
[2021-07-28T06:56:33.408Z] github.com/pingcap/tidb/store/tikv.(*RegionRequestSender).SendReq(0xc000e0c280, 0xc000156400, 0xc0021e6b60, 0x3, 0x0, 0x0, 0x4a817c800, 0x0, 0x0, 0x0)
[2021-07-28T06:56:33.408Z] 	/nfs/cache/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20210330094614-60111e1c4b6f/store/tikv/region_request.go:186 +0x17e
[2021-07-28T06:56:33.408Z] github.com/pingcap/tidb/store/tikv.actionPrewrite.handleSingleBatch(0xc000df4000, 0xc000156400, 0x3, 0x0, 0x0, 0x8511108, 0xc00143c160, 0x1, 0x0, 0x0)
[2021-07-28T06:56:33.408Z] 	/nfs/cache/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20210330094614-60111e1c4b6f/store/tikv/prewrite.go:162 +0x815
[2021-07-28T06:56:33.408Z] github.com/pingcap/tidb/store/tikv.(*twoPhaseCommitter).doActionOnBatches(0xc000df4000, 0xc000156400, 0x84f2f20, 0xabbedf8, 0xc0030bc7e0, 0x1, 0x1, 0x0, 0x0)
[2021-07-28T06:56:33.408Z] 	/nfs/cache/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20210330094614-60111e1c4b6f/store/tikv/2pc.go:649 +0x230
[2021-07-28T06:56:33.408Z] github.com/pingcap/tidb/store/tikv.(*twoPhaseCommitter).doActionOnGroupMutations(0xc000df4000, 0xc000156400, 0x84f2f20, 0xabbedf8, 0xc0030bc7b0, 0x1, 0x1, 0x0, 0x0)
[2021-07-28T06:56:33.408Z] 	/nfs/cache/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20210330094614-60111e1c4b6f/store/tikv/2pc.go:630 +0x12e5
[2021-07-28T06:56:33.408Z] github.com/pingcap/tidb/store/tikv.(*twoPhaseCommitter).doActionOnMutations(0xc000df4000, 0xc000156400, 0x84f2f20, 0xabbedf8, 0x8511108, 0xc00143c100, 0x0, 0x0)
[2021-07-28T06:56:33.408Z] 	/nfs/cache/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20210330094614-60111e1c4b6f/store/tikv/2pc.go:493 +0x351
[2021-07-28T06:56:33.408Z] github.com/pingcap/tidb/store/tikv.(*twoPhaseCommitter).prewriteMutations(0xc000df4000, 0xc000156400, 0x8511108, 0xc00143c100, 0x0, 0x0)
[2021-07-28T06:56:33.408Z] 	/nfs/cache/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20210330094614-60111e1c4b6f/store/tikv/prewrite.go:289 +0x55f
[2021-07-28T06:56:33.408Z] github.com/pingcap/tidb/store/tikv.(*twoPhaseCommitter).execute(0xc000df4000, 0x84f9d08, 0xc00021e010, 0x0, 0x0)
[2021-07-28T06:56:33.408Z] 	/nfs/cache/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20210330094614-60111e1c4b6f/store/tikv/2pc.go:1020 +0x970
[2021-07-28T06:56:33.408Z] github.com/pingcap/tidb/store/tikv.(*KVTxn).Commit(0xc0001b6000, 0x84f9d08, 0xc00021e010, 0x0, 0x0)
[2021-07-28T06:56:33.408Z] 	/nfs/cache/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20210330094614-60111e1c4b6f/store/tikv/txn.go:279 +0xb72
[2021-07-28T06:56:33.408Z] github.com/pingcap/tidb/store/driver/txn.(*tikvTxn).Commit(0xc002ee22b0, 0x84f9d08, 0xc00021e010, 0x0, 0x0)
[2021-07-28T06:56:33.408Z] 	/nfs/cache/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20210330094614-60111e1c4b6f/store/driver/txn/txn_driver.go:61 +0xa7
[2021-07-28T06:56:33.408Z] github.com/pingcap/tidb/kv.RunInNewTxn(0x84f9d08, 0xc00021e010, 0x8521950, 0xc000fbf320, 0x1, 0xc0018de658, 0x0, 0x0)
[2021-07-28T06:56:33.408Z] 	/nfs/cache/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20210330094614-60111e1c4b6f/kv/txn.go:60 +0xa62
[2021-07-28T06:56:33.408Z] github.com/pingcap/tidb/ddl.(*ddl).genGlobalIDs(0xc000fe8000, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
[2021-07-28T06:56:33.408Z] 	/nfs/cache/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20210330094614-60111e1c4b6f/ddl/ddl.go:421 +0x228
[2021-07-28T06:56:33.408Z] github.com/pingcap/tidb/ddl.(*ddl).CreateSchemaWithInfo(0xc000fe8000, 0x852c170, 0xc001ea9400, 0xc001aae070, 0x0, 0x0, 0x0)
[2021-07-28T06:56:33.408Z] 	/nfs/cache/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20210330094614-60111e1c4b6f/ddl/ddl_api.go:121 +0x6f9
[2021-07-28T06:56:33.408Z] github.com/pingcap/tidb/ddl.(*ddl).CreateSchema(0xc000fe8000, 0x852c170, 0xc001ea9400, 0xc0018fb788, 0x7, 0xc0018fb788, 0x7, 0x0, 0x0, 0x0)
[2021-07-28T06:56:33.408Z] 	/nfs/cache/mod/github.com/pingcap/tidb@v1.1.0-beta.0.20210330094614-60111e1c4b6f/ddl/ddl_api.go:89 +0x5ef
[2021-07-28T06:56:33.408Z] github.com/pingcap/dm/pkg/schema.(*Tracker).CreateSchemaIfNotExists(0xc001953650, 0xc0018fb788, 0x7, 0x0, 0x0)
[2021-07-28T06:56:33.408Z] 	/home/jenkins/agent/workspace/dm_ghpr_new_test/go/src/github.com/pingcap/dm/pkg/schema/tracker.go:290 +0x3f2
[2021-07-28T06:56:33.408Z] github.com/pingcap/dm/syncer.(*RemoteCheckPoint).Rollback(0xc00088f950, 0xc001953650)
[2021-07-28T06:56:33.408Z] 	/home/jenkins/agent/workspace/dm_ghpr_new_test/go/src/github.com/pingcap/dm/syncer/checkpoint.go:701 +0x1146
[2021-07-28T06:56:33.408Z] github.com/pingcap/dm/syncer.(*Syncer).Process(0xc000203000, 0x84f9cd0, 0xc0019ef080, 0xc002135ce0)
[2021-07-28T06:56:33.408Z] 	/home/jenkins/agent/workspace/dm_ghpr_new_test/go/src/github.com/pingcap/dm/syncer/syncer.go:678 +0xfa8
[2021-07-28T06:56:33.408Z] created by github.com/pingcap/dm/dm/worker.(*SubTask).run
[2021-07-28T06:56:33.408Z] 	/home/jenkins/agent/workspace/dm_ghpr_new_test/go/src/github.com/pingcap/dm/dm/worker/subtask.go:230 +0xbcb
```

File: syncer/syncer.go [hunk 3]
```diff
@@ -1590,7 +1584,11 @@ func (s *Syncer) Run(ctx context.Context) (err error) {
 		} else {
 			err2 = s.flushCheckPoints()
 		}
-		tctx.L().Warn("fail to flush check points when exit task", zap.Error(err2))
+		if err != nil {
```
Comment (lichunzhu) [inline syncer/syncer.go:1587]: Do you mean this?
```suggestion
		if err2 != nil {
```

File: syncer/syncer.go [hunk 4]
```diff
@@ -1554,6 +1572,7 @@ func (s *Syncer) Run(ctx context.Context) (err error) {
 			err2            error
 			exitSafeModeLoc binlog.Location
 		)
+		// TODO: why use a sharding-resync variable savedGlobalLastLocation?
```
Comment (lichunzhu) [inline syncer/syncer.go:1575]: To make safe more `safe`.
pos1 ... pos2 ... pos3
If DM resync between pos1 and pos2, I want safe mode open till pos3

File: syncer/syncer.go [hunk 5]
```diff
@@ -1565,7 +1584,11 @@ func (s *Syncer) Run(ctx context.Context) (err error) {
 		} else {
 			err2 = s.flushCheckPoints()
 		}
-		tctx.L().Warn("fail to flush check points when exit task", zap.Error(err2))
+		if err != nil {
```
Comment (lichunzhu) [inline syncer/syncer.go:1587]: Do you mean `err2` here?

File: syncer/syncer.go [hunk 6]
```diff
@@ -1781,7 +1804,9 @@ func (s *Syncer) Run(ctx context.Context) (err error) {
 		// check pass SafeModeExitLoc and try disable safe mode, but not in sharding or replacing error
 		safeModeExitLoc := s.checkpoint.SafeModeExitPoint()
 		if safeModeExitLoc != nil && !s.isReplacingErr && shardingReSync == nil {
-			if binlog.CompareLocation(currentLocation, *safeModeExitLoc, s.cfg.EnableGTID) >= 0 {
+			// TODO: for RowsEvent (in fact other than QueryEvent), `currentLocation` is updated in `handleRowsEvent`
```
Comment (lichunzhu) [inline syncer/syncer.go:1807]: Will this cause a problem here?

**Review by glorv — APPROVED** [review]

File: syncer/syncer.go (hunk 6 above)
Comment (lance6716) [reply syncer/syncer.go:1807]: originally all `currentLocation` are located in `handleXXX`, to support handle-error, some DDL injections are needed in main routine so for query event `currentLocation` is updated earlier. cc @GMHDBJD 

I'll fix above comments when this CI finishes

Comment (lance6716) [reply syncer/syncer.go:1806]: CI success

Comment (lichunzhu) [reply syncer/syncer.go:1806]: I have thought this over again. If we update `currentLocation` later after `XIDEvent`, we will open a longer safemode (one transaction). But it won't affect the correctness.

**Comment by lance6716** [comment]
/run-all-tests

an irrelevant UT failed

**Comment by lance6716** [comment]
@lichunzhu

**Review by lichunzhu — APPROVED** [review]
LGTM

**Comment by Ehco1996** [comment]
> /run-all-tests
> 
> an irrelevant UT failed

/hold 

oh  i find why
```
[2021-07-28T23:26:16.013Z] [2021/07/29 07:26:02.351 +08:00] [DEBUG] [syncer.go:1382] ["no job in queue, update lag to zero"] [task=test] [unit="binlog replication"] [workerLagKey=lag_0] ["current ts"=1627514762]
[2021-07-28T23:26:16.013Z] [2021/07/29 07:26:02.437 +08:00] [DEBUG] [syncer.go:712] ["lazy init table info in schema tracker"] [task=test] [unit="binlog replication"] [schema=metrics] [table=t1]
[2021-07-28T23:26:16.013Z] [2021/07/29 07:26:02.438 +08:00] [DEBUG] [syncer.go:2707] ["key for keys"] [task=test] [unit="binlog replication"] [key=1004`metrics`.`t1`] [keys="[1004`metrics`.`t1`]"]
[2021-07-28T23:26:16.013Z] [2021/07/29 07:26:02.438 +08:00] [DEBUG] [syncer.go:980] ["queue for key"] [task=test] [unit="binlog replication"] [queue=0] [key=1004`metrics`.`t1`]
[2021-07-28T23:26:16.013Z] [2021/07/29 07:26:02.439 +08:00] [DEBUG] [checkpoint.go:375] ["save table checkpoint"] [task=test] [unit="binlog replication"] [component="remote checkpoint"] [location="position: (dm-integration-test-qkf5b-mgpdc-bin.000001, 2268), gtid-set: ac76b687-eff9-11eb-ba46-72fb517c9433:1-9"] [schema=metrics] [table=t1]
[2021-07-28T23:26:16.013Z] [2021/07/29 07:26:02.440 +08:00] [DEBUG] [syncer.go:1745] ["receive binlog event"] [task=test] [unit="binlog replication"] [header="{\"Timestamp\":1627514762,\"EventType\":16,\"ServerID\":1,\"EventSize\":31,\"LogPos\":2299,\"Flags\":0}"]
[2021-07-28T23:26:16.013Z] [2021/07/29 07:26:02.440 +08:00] [DEBUG] [baseconn.go:164] ["execute statement"] [task=test] [unit="binlog replication"] [query="INSERT INTO `metrics`.`t1` (`id`,`name`,`dt`,`ts`,`new_col1`) VALUES (?,?,?,?,?)"] [argument="[1004 zmj4 <nil> 2022-05-11 12:01:05 <nil>]"]
[2021-07-28T23:26:16.013Z] [2021/07/29 07:26:02.440 +08:00] [DEBUG] [syncer.go:1876] [task=test] [unit="binlog replication"] [event=XID] ["last location"="position: (dm-integration-test-qkf5b-mgpdc-bin.000001, 2163), gtid-set: ac76b687-eff9-11eb-ba46-72fb517c9433:1-9"] [location="position: (dm-integration-test-qkf5b-mgpdc-bin.000001, 2299), gtid-set: ac76b687-eff9-11eb-ba46-72fb517c9433:1-9"]
[2021-07-28T23:26:16.013Z] [2021/07/29 07:26:02.440 +08:00] [DEBUG] [checkpoint.go:491] ["save global checkpoint"] [task=test] [unit="binlog replication"] [component="remote checkpoint"] [location="position: (dm-integration-test-qkf5b-mgpdc-bin.000001, 2299), gtid-set: ac76b687-eff9-11eb-ba46-72fb517c9433:1-9"]
[2021-07-28T23:26:16.013Z] [2021/07/29 07:26:03.601 +08:00] [DEBUG] [server.go:284] ["etcd member list doesn't change"] ["client URLs"="[http://127.0.0.1:8261]"]
[2021-07-28T23:26:16.013Z] [2021/07/29 07:26:03.698 +08:00] [DEBUG] [worker.go:146] ["runtime status"] [component="worker controller"] [status="{\n    \"status\": [\n        {\n            \"name\": \"test\",\n            \"stage\": \"Running\",\n            \"unit\": \"Sync\",\n            \"result\": null,\n            \"unresolvedDDLLockID\": \"\",\n            \"sync\": {\n                \"totalEvents\": \"1\",\n                \"totalTps\": \"0\",\n                \"recentTps\": \"0\",\n                \"masterBinlog\": \"\",\n                \"masterBinlogGtid\": \"\",\n                \"syncerBinlog\": \"(dm-integration-test-qkf5b-mgpdc-bin.000001, 2015)\",\n                \"syncerBinlogGtid\": \"ac76b687-eff9-11eb-ba46-72fb517c9433:1-8\",\n                \"blockingDDLs\": [\n                ],\n                \"unresolvedGroups\": [\n                ],\n                \"synced\": false,\n                \"binlogType\": \"remote\",\n                \"secondsBehindMaster\": \"0\"\n            }\n        }\n    ]\n}"]
```

after one dml executed , syncer started to `save global checkpoint`, and it spend more than 5s 😂
let me fix it

**Comment by Ehco1996** [comment]
@lance6716 CI passed PTAL

**Comment by lance6716** [comment]
@Ehco1996 please unhold if you think it's ok

**Pull request merged**